Add quiescence negotiation

remyers opened the pull request:
This draft is to get feedback on the overall design; if it looks like the right direction I will move the changes needed to enable a splice while quiescent to another PR. This one should only include what is necessary for quiescence negotiation up to the point  of starting the splice.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/Features.scala:
```diff
@@ -300,6 +300,13 @@ object Features {
+  // TODO: @remyers option_quiesce implementation, to be replaced once quiescence is spec-ed
+  case object QuiescePrototype extends Feature with InitFeature {
+    val rfcName = "option_quiesce_prototype"
+    val mandatory = 158
```
Why don't you use the real feature bit from the spec PR? I don't think see a good reason to use a temporary one...

remyers replied:
The [quiescence spec PR](https://github.com/lightning/bolts/pull/869) uses 34/35, but that is not listed in the [waiting room](https://github.com/lightning/bolts/issues/605). In fact the waiting room uses 34/35 for 'option_inv_gossip'.

Should I still use 34/35 or work this out first on the lightning-dev call and then update the waiting room?

remyers replied:
otoh - we can assume the waiting room is out of date, so I took your suggestion to use the spec bits.

Fixed in [9fa905a](https://github.com/ACINQ/eclair/pull/2680/commits/9fa905a3e557420b011b8c5079cb48ba74b705a2)

t-bast replied:
Yes, spec PRs always have precedence, you should just ignore the waiting room (which shouldn't use low feature bits in the first place).

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala:
```diff
@@ -764,7 +764,7 @@ case class Commitments(params: ChannelParams,
   // @formatter:off
   // HTLCs and pending changes are the same for all active commitments, so we don't need to loop through all of them.
-  def isIdle: Boolean = active.head.isIdle(changes)
+  def isIdle: Boolean = active.head.isIdle(changes, params.remoteParams.initFeatures.hasFeature(QuiescePrototype))
```
Nice, good idea to use the feature bits instead of exposing the boolean. I think we should check both local and remote features though, we apply quiescence only if both nodes advertise it.

remyers replied:
Fixed in [9fa905a](https://github.com/ACINQ/eclair/pull/2680/commits/9fa905a3e557420b011b8c5079cb48ba74b705a2)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelData.scala:
```diff
@@ -452,8 +451,13 @@ object RbfStatus {
+sealed trait QuiescenceNegotiation extends SpliceStatus
 object SpliceStatus {
   case object NoSplice extends SpliceStatus
+  case class QuiescenceRequested(splice: CMD_SPLICE) extends QuiescenceNegotiation
+  case class InitiatorQuiescent(splice: CMD_SPLICE) extends QuiescenceNegotiation
+  case class ReceivedStfu(stfu: Stfu) extends QuiescenceNegotiation
+  case object NonInitiatorQuiescent extends QuiescenceNegotiation
```
Just to make sure I understand that correctly, there are two distinct paths in the state machine depending on whether we're the initiator or not:

- if we're the initiator: `QuiescenceRequested -> InitiatorQuiescent -> NonInitiatorQuiescent`
- if we're not the initiator: `ReceivedStfu -> NonInitiatorQuiescent`

The non-initiator always sends its `stfu` after all of its pending updates have been signed, right? So when we receive `stfu`, we should verify that the sending side doesn't have any pending update. What about renaming `NonInitiatorQuiescent` to `QuiescenceReached` instead? Because at that point all updates in both directions have been signed?

remyers replied:
I had previously named `NonInitiatorQuiescent` to `Quiescent` (equivalent to `QuiescenceReached` but now the state machine goes like this:
 `QuiescenceRequested -> InitiatorQuiescent -> SpliceRequested` 

The last transition happens when the initiator receives `stfu` from the non-initiator.

t-bast replied:
Right, that looks correct! It's even simpler than what I had in mind, this is better :+1:

t-bast commented on eclair-core/src/main/resources/reference.conf:
```diff
@@ -151,6 +151,8 @@ eclair {
       channel-opener-whitelist = [] // a list of public keys; we will ignore rate limits on pending channels from these peers
     }
+
+    quiescence-timeout = 2 minutes // maximum time we will wait for quiescence to complete before disconnecting
```
nit: 1 minute is probably more than enough here, we're just exchanging `commit_sig` / `revoke_and_ack` in both directions and nothing more.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageCodecs.scala:
```diff
@@ -476,6 +480,7 @@ object LightningMessageCodecs {
     .typecase(37000, spliceInitCodec)
     .typecase(37002, spliceAckCodec)
     .typecase(37004, spliceLockedCodec)
+    .typecase(37006, stfuCodec)
```
Same here, no need to use a custom message type, we should directly use the one from the spec PR.

remyers replied:
Rusty chose `2`? I so we're now back filling the early bits?

Fixed in [9fa905a](https://github.com/ACINQ/eclair/pull/2680/commits/9fa905a3e557420b011b8c5079cb48ba74b705a2)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageTypes.scala:
```diff
@@ -560,6 +560,7 @@ case class GossipTimestampFilter(chainHash: ByteVector32, firstTimestamp: Timest
 
 case class OnionMessage(blindingKey: PublicKey, onionRoutingPacket: OnionRoutingPacket, tlvStream: TlvStream[OnionMessageTlv] = TlvStream.empty) extends LightningMessage
 
+case class Stfu(channelId: ByteVector32, initiator: Byte) extends LightningMessage
```
nit: should be right before `SpliceInit`

remyers replied:
Fixed in [9fa905a](https://github.com/ACINQ/eclair/pull/2680/commits/9fa905a3e557420b011b8c5079cb48ba74b705a2)

t-bast reviewed (commented):
It looks like this is going in the right direction, it's ready to be cleaned up to only contain the quiescence steps and then we'll be able to do a more thorough review.

Note that you're not currently targeting the `master` branch, but rather an obsolete splice branch (`splices-bast-pm`), make sure to rebase on `master` for the "official" PR.

I think that the way we should handle that first PR without impacting the splice prototype is that whenever `CMD_SPLICE` or `SpliceInit` is requested, we should look at our _local_ feature bits: if `quiescence` is unable, we go through the quiescence steps, otherwise we skip it (and use `isIdle` without htlcs).

remyers commented:
I  moved the (in-progress) changes to splice to accommodate commitment txs with htlcs to [splice-with-htlcs](https://github.com/remyers/eclair/tree/splice-with-htlcs) pending this PR.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageTypes.scala:
```diff
@@ -278,6 +278,8 @@ case class ChannelReady(channelId: ByteVector32,
   val alias_opt: Option[Alias] = tlvStream.get[ShortChannelIdTlv].map(_.alias)
 }
 
+case class Stfu(channelId: ByteVector32, initiator: Byte) extends LightningMessage
```
We should use a `Boolean` here instead of a byte: see https://github.com/ACINQ/eclair/commit/e085b02359c0b692f199415a24064018f4fa2b80

remyers replied:
Fixed in [ab9cf21](https://github.com/ACINQ/eclair/pull/2680/commits/ab9cf211bade60f52c78971d771b983fff925e5b)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:
```diff
@@ -363,15 +367,37 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
       c match {
         case c: CMD_ADD_HTLC => handleAddHtlcCommandError(c, error, Some(d.channelUpdate))
-        // NB: the command cannot be an htlc settlement (fail/fulfill), because if we are splicing it means the channel is idle and has no htlcs
+        // NB: the command cannot be an htlc settlement (fail/fulfill), because if we are splicing without quiescence support it means the channel is idle and has no htlcs
```
Actually at that point we can perfectly well be using quiescence (and have already reached the quiescent state), so the comment is incorrect. But more fundamentally, I think we should include the splicing protocol inside the quiescence timeout timer: we don't want to stay quiescent too long, because it creates a risk of pending HTLCs timing out and it makes the channel unable to forward HTLCs (and thus loose fee revenue). If either peer has a bug that makes the splicing protocol block, we want to disconnect to restore the channel to a known good state.

I've implemented this in https://github.com/ACINQ/eclair/commit/2476fb83a3440891fc6b24fbd76bcdb6e82b8c71

remyers replied:
Good idea to extend the quiescence time out period to include the splice negotiation. Fixed in [5841138](https://github.com/ACINQ/eclair/pull/2680/commits/58411385f7da40e50f5180a7185344317b608b73).

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:
```diff
@@ -781,51 +808,62 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
-            log.warning("cannot initiate splice, channel is not idle or peer doesn't support splices")
-            cmd.replyTo ! RES_FAILURE(cmd, CommandUnavailableInThisState(d.channelId, "splice", NORMAL))
+          case SpliceStatus.QuiescenceRequested(_) =>
+            stay() using d.copy(spliceStatus = SpliceStatus.ReceivedStfu(msg))
```
This case is a red herring that we're not handling correctly the race between our `stfu` and theirs, because we're switching from initiator to non-initiator implicitly here (which would create a deadlock if we were the channel initiator).

But there is an issue in the data model, which currently doesn't let us do that easily, because if we're the channel initiator and stay in `QuiescenceRequested`, we lose the information that our peer has already sent us `stfu` and will deadlock later waiting for them to send it.

We probably need to add a field `remoteStfuReceived: Boolean` to `QuiescenceRequested` to handle that case?

remyers replied:
I don't see how we can get a deadlock here. There should be no race possible between our `stfu` and theirs in the `SpliceStatus.QuiescenceRequested` state because we have not sent an `stfu` yet. This state means we are simply waiting for our local pending updates to finalize before sending an initial `stfu`.

If a node in this state receives `stfu` from it's peer before it has sent it's own `stfu`, then it's safe to switch to being the non-initiator because we have not sent our own `stfu` yet.

t-bast replied:
Good catch, you're right, this works fine!

t-bast commented on eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalQuiescentStateSpec.scala:
```diff
@@ -0,0 +1,434 @@
+    receiveSettlementCommand(f, FailHtlc, sendInitialStfu = false)
+  }
+
+  test("recv fail malformed htlc command while initiator awaiting stfu from remote") { f =>
```
It's honestly overkill to duplicate tests for fail and fail_malformed (both for commands and messages), I think you can remove the malformed ones.

remyers replied:
Fixed in [1910599](https://github.com/ACINQ/eclair/pull/2680/commits/1910599539cf2852a6b3068b3fbb5d72b49dd020)

t-bast reviewed (commented):
Great work, this is a very subtle part of the codebase but you managed to figure it out on your own, with most of it working correctly!

I've made a few small changes in [this branch](https://github.com/ACINQ/eclair/commits/quiescence), have a look at each commit separately to see if it makes sense and then feel free to cherry-pick them before making additional changes (it makes the overall diff with `master` smaller).

There are some test scenarios that I'd like to add that are currently missing:

- remote sends `stfu` but isn't quiescent -> verify we disconnect when we reach the `splice_init` step (test the case where remote is initiator and the non-initiator case)
- concurrent `stfu` (we should have a test for the deadlock that exists currently in the implementation - see comment below)
- a received HTLC times out while we're waiting for quiescence, and we have the preimage in our pending commands -> verify that we force-close and extract the preimage to publish HTLC-success
- quiescence timeout reached in the middle of the splicing part of the protocol: don't test every sub-state, that would be overkill, but it's worth testing the following two cases:
  - timeout while in `SpliceInProgress`
  - timeout while in `SpliceAborted` when we don't receive our peer's `tx_abort` (I'm quite afraid of bugs where peers won't correctly echo `tx_abort` and the channel will be stuck, but it's now fixed by the quiescence timeout that will force a reconnection)

t-bast commented:
Also, this needs a rebase ;)

remyers commented:
The behavior of `splice-init` **without quiescence**  has changed and is now reflected in the test change in [bc59999](https://github.com/ACINQ/eclair/pull/2680/commits/bc5999941412cf105df5cd95d2b29a04b316e4fa) when there is a race between the `splice-init` from Alice and an `add-htlc` from Bob. Before the `add-htlc` would have gone through and the splice would have been aborted. This situation now triggers a disconnect from Alice and the `add-htlc` does not go through.

Is this the desired behavior? It's more consistent with how `splice-init` **with quiescence** works, but could lead to unnecessary disconnects and failed `add-htlc` when quiescence is not negotiated.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:
```diff
@@ -781,51 +808,62 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
+            }
+          case SpliceStatus.ReceivedStfu(_) | SpliceStatus.NonInitiatorQuiescent =>
+            cancelTimer(QuiescenceTimeout.toString)
+            val failure = new ChannelException(d.channelId, "received stfu twice")
```
Good point; fixed in [9225040](https://github.com/ACINQ/eclair/pull/2680/commits/9225040ef2b028ebb104e82baa435b9ce867e4ca)

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:
```diff
@@ -363,15 +367,37 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
       }
 
-    case Event(msg: ForbiddenMessageDuringSplice, d: DATA_NORMAL) if d.spliceStatus != SpliceStatus.NoSplice && !d.spliceStatus.isInstanceOf[SpliceStatus.SpliceRequested] =>
+    case Event(msg: ForbiddenMessageDuringSplice, d: DATA_NORMAL) if d.commitments.params.useQuiescence && d.spliceStatus != SpliceStatus.NoSplice && !d.spliceStatus.isInstanceOf[SpliceStatus.InitiatorQuiescent] =>
```
In change [1e7dfad](https://github.com/ACINQ/eclair/pull/2680/commits/1e7dfad4e6287a69843ea0efdb9e568b2f3a48be) I changed `NonInitiatorQuiescent` to extend `QuiescentSpliceStatus` instead of `QuiescenceNegotiation.NonInitiator`. Logically once the non-initiator sends `stfu` they are quiescent (no longer negotiating quiescence), even if the initiator might not have yet received their `stfu`. This change allowed me to simplify the tests for forbidden messages that I previously called "... after initiator receives stfu from remote" and rename them to "... while quiescent".

Because forbidden messages do not trigger a channel close, I could simplify the tests to ensure both alice and bob return a warning (and disconnect) in the same test and remove the separate tests for bob. 

NB: while renaming the tests, I found/removed a duplicate test called "recv settlement commands after initiator receives stfu from remote" which was identical to "recv forbidden commands while quiescent".

remyers commented:
I've added the tests you recommend:

> * remote sends `stfu` but isn't quiescent -> verify we disconnect when we reach the `splice_init` step (test the case where remote is initiator and the non-initiator case)

In [a7875ce](https://github.com/ACINQ/eclair/pull/2680/commits/a7875ce1e6563878480d97be3c1dcb1d1d0ec35c) I detect this immediately when `stfu` is received from a remote that is not quiescent; I also added tests for both initiator/non-initiator. 

> * concurrent `stfu` (we should have a test for the deadlock that exists currently in the implementation - see comment below)

There should not be a deadlock (see [comment above](https://github.com/ACINQ/eclair/pull/2680#discussion_r1269160772)), but still a good idea to add a test for concurrent `stfu` in [ea5e3a1](https://github.com/ACINQ/eclair/pull/2680/commits/ea5e3a1bd82b3e17287c5da32cd6780248a0fd70).

> * a received HTLC times out while we're waiting for quiescence, and we have the preimage in our pending commands -> verify that we force-close and extract the preimage to publish HTLC-success

In [9b1864c](https://github.com/ACINQ/eclair/pull/2680/commits/9b1864cf0509d73e9f95cc1e9793462c556ac293) I created test "htlc timeout during quiescence negotiation with pending preimage" but it did not behave as I expected. There is something going on that I don't understand related to the commitment mechanism. Any idea why the `htlcSuccess` tx is not published when the `initialCommittx` is first published? 

> * quiescence timeout reached in the middle of the splicing part of the protocol: don't test every sub-state, that would be overkill, but it's worth testing the following two cases:
>   * timeout while in `SpliceInProgress`
>   * timeout while in `SpliceAborted` when we don't receive our peer's `tx_abort` (I'm quite afraid of bugs where peers won't correctly echo `tx_abort` and the channel will be stuck, but it's now fixed by the quiescence timeout that will force a reconnection)

I added timeout tests for these states in [9b1864c](https://github.com/ACINQ/eclair/pull/2680/commits/9b1864cf0509d73e9f95cc1e9793462c556ac293).

Question: should I test for the disconnect in both timeout tests? Is it enough to just check for `InvalidSpliceNotQuiescent` in one of the two tests?

remyers commented:
> Great work, this is a very subtle part of the codebase but you managed to figure it out on your own, with most of it working correctly!

Thanks! It was more complicated then I initially thought (as always)/ My [notes](https://gist.github.com/remyers/5f556b7063fc06cc2b3cd10f69169c89) might help someone in the future understand it too, even if they are a bit out of date now.

remyers commented:
> Also, this needs a rebase ;)

Rebased and ready for review!

t-bast commented:
> The behavior of splice-init without quiescence has changed and is now reflected in the test change in [d250f39](https://github.com/ACINQ/eclair/pull/2680/commits/d250f39e88fe2b3f62d3706f0969560b58525c7c) when there is a race between the splice-init from Alice and an add-htlc from Bob. Before the add-htlc would have gone through and the splice would have been aborted. This situation now triggers a disconnect from Alice and the add-htlc does not go through.
>
> Is this the desired behavior? It's more consistent with how splice-init with quiescence works, but could lead to unnecessary disconnects and failed add-htlc when quiescence is not negotiated.

Yes, I think this is ok, because:

- this is an edge case that should be rare 
- we'll want to start using quiescence and deprecate the non-quiescent splices whenever possible
- this is still providing a way out of the edge case that doesn't create a force-close

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:
```diff
@@ -784,54 +825,84 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
+              } else {
+                stay() using d.copy(spliceStatus = SpliceStatus.ReceivedStfu(msg))
+              }
+            case SpliceStatus.QuiescenceRequested(_) =>
+              stay() using d.copy(spliceStatus = SpliceStatus.ReceivedStfu(msg))
```
We need to report the splice failure, and it's also worth logging something:

```suggestion
            case SpliceStatus.QuiescenceRequested(cmd) =>
              // We could keep track of our splice attempt and merge it with the remote splice instead of cancelling it.
              // But this is an edge case that should rarely occur, so it's probably not worth the additional complexity.
              log.warning("our peer initiated quiescence before us, cancelling our splice attempt")
              cmd.replyTo ! RES_FAILURE(cmd, ConcurrentRemoteSplice(d.channelId))
              stay() using d.copy(spliceStatus = SpliceStatus.ReceivedStfu(msg))
```

Where I added this error after `InvalidSpliceNotQuiescent` in `ChannelExceptions.scala`:

```scala
case class ConcurrentRemoteSplice                  (override val channelId: ByteVector32) extends ChannelException(channelId, "cancelling splice attempt: remote is also trying to splice")
```

remyers replied:
Fixed in [a31a96f](https://github.com/ACINQ/eclair/pull/2680/commits/a31a96f3d756fdc8919fae6c363cf90757047b14).

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:
```diff
@@ -784,54 +825,84 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
+                    stay() using d.copy(spliceStatus = SpliceStatus.SpliceRequested(cmd, spliceInit)) sending spliceInit
+                }
+              } else {
+                stay() using d.copy(spliceStatus = SpliceStatus.NonInitiatorQuiescent)
```
We should also report the failure and log something here, I'm curious to see if this race condition happens frequently in practice:

```suggestion
                log.warning("concurrent stfu received and our peer is the channel initiator, cancelling our splice attempt")
                cmd.replyTo ! RES_FAILURE(cmd, ConcurrentRemoteSplice(d.channelId))
                stay() using d.copy(spliceStatus = SpliceStatus.NonInitiatorQuiescent)
```

remyers replied:
fixed in [a31a96f](https://github.com/ACINQ/eclair/pull/2680/commits/a31a96f3d756fdc8919fae6c363cf90757047b14).

t-bast reviewed (commented):
The code looks good to me! I still need to review the tests, but otherwise I couldn't find any issue.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelData.scala:
```diff
@@ -450,12 +451,32 @@ object RbfStatus {
+  case class InitiatorQuiescent(splice: CMD_SPLICE) extends QuiescenceNegotiation.Initiator
+  /** Our peer is asked us to stop sending new updates and wait for our updates to be added to the local and remote commitments. */
+  case class ReceivedStfu(stfu: Stfu) extends QuiescenceNegotiation.NonInitiator
+  /** Our updates to be added to the local and remote commitments, we wait for our peer to use the now quiescent channel. */
```
fixed in [316f770](https://github.com/ACINQ/eclair/pull/2680/commits/316f770424af684dd9ae04ec17cebc6fb3efb645)

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelData.scala:
```diff
@@ -450,12 +451,32 @@ object RbfStatus {
+  case class QuiescenceRequested(splice: CMD_SPLICE) extends QuiescenceNegotiation.Initiator
+  /** Our updates to be added to the local and remote commitments, we wait for our peer to do the same. */
+  case class InitiatorQuiescent(splice: CMD_SPLICE) extends QuiescenceNegotiation.Initiator
+  /** Our peer is asked us to stop sending new updates and wait for our updates to be added to the local and remote commitments. */
```
I should have caught that too! fixed in [316f770](https://github.com/ACINQ/eclair/pull/2680/commits/316f770424af684dd9ae04ec17cebc6fb3efb645)

t-bast reviewed (commented):
I've added several commits on top of your PR in https://github.com/ACINQ/eclair/tree/quiescence-bast, can you have a look at them and cherry-pick?

I'm mostly simplifying some tests, adding others, and moving stuff around to keep them grouped by "feature".

Regarding the force-close test, it was behaving correctly: when we detect that we need to force-close, we immediately generate the corresponding transactions and broadcast them. At that point, we only look at our local data, so we don't notice that we have the preimage. Then when transitioning to the `CLOSING` state (look for the corresponding `onTransition`), we check the pending commands DB and replay the `CMD_FULFILL_HTLC`, which makes us generate a new set of force-closing transactions that include the HTLC-success transaction.

remyers commented on eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala:
```diff
@@ -649,21 +649,20 @@ class NormalSplicesStateSpec extends TestKitBaseClass with FixtureAnyFunSuiteLik
-    crossSign(bob, alice, bob2alice, alice2bob)
-    fulfillHtlc(add.id, preimage, alice, bob, alice2bob, bob2alice)
-    crossSign(alice, bob, alice2bob, bob2alice)
+    // the htlc is not added
```
fixed in [7082c4a](https://github.com/ACINQ/eclair/pull/2680/commits/7082c4ae7c71d02c10a5451881243bacfd36064a)

remyers commented:
> I've added several commits on top of your PR in https://github.com/ACINQ/eclair/tree/quiescence-bast, can you have a look at them and cherry-pick?

Thanks! I've cherry-picked those changes. I only added [f51710e](https://github.com/ACINQ/eclair/pull/2680/commits/f51710e427b850cc23344ac5a620b2871e1d1081) to check that one side receives a command failure during concurrent splices.

I also made the change you suggested to `NormalSplicesStateSpec` to confirm an htlc was not added: [7082c4a](https://github.com/ACINQ/eclair/pull/2680/commits/7082c4ae7c71d02c10a5451881243bacfd36064a).

t-bast reviewed (approved):
LGTM, one more step towards a full splicing implementation! :rocket:
